core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -537,14 +533,14 @@ class GroupCoordinator(val brokerId: Int,
           // TODO: cut the socket connection to the client
         }
 
+        // TODO KAFKA-2720: only remove group in the background thread
         if (group.isEmpty) {
-          group.transitionTo(Dead)
+          groupManager.removeGroup(group)
```

Comment:
I think the synchronization around adding/removing groups from the cache might not be sufficient. Suppose that this removeGroup() is hit while another thread is blocking on the group lock in doJoinGroup(). The group will be removed from the cache, but the other thread will still be able to acquire the lock when it is finished and perform operations on the group, right?